### DISCUSSION TITLE
how to use pybamm.diff()

### DISCUSSION THREAD

**Discussion by xrfe** [pybamm-team/PyBaMM#3420] [Q&A] [answered]
when I create my submodel, one of the term on the right hand side depends on the time derivative, I then use pybamm.diff(), 
e.g., dcdt = K*pybamm.diff(j_n, pybamm.t)
but got error: 

`    return self.left.diff(variable) + self.right.diff(variable)

  File /pybamm/lib/python3.11/site-packages/pybamm/expression_tree/unary_operators.py:345 in diff
    raise NotImplementedError

NotImplementedError`

Does anyone know how to solve the issue?
Thanks!

**Comment by brosaplanella** [comment]
I suspect this is to do with how `j_n` is defined. Can you provide a minimum working example?

**Reply by xrfe** [reply]
Thanks for the reply. 
sorry for confision, my problem is
when creating models, how to implement time-derivatives of internal variables of pybamm, such as current density (j_n----d j_n / dt), overpotential (η ----- dη/dt)?
for example,  (https://github.com/pybamm-team/PyBaMM/blob/develop/examples/scripts/create_model.py):
in Line 41:

`N = -(1 / L) * D(c) * pybamm.grad(c)  # solvent flux,`

now if N depends on the derivative of any intermediate varialbles, e.g., 
N = -(1 / L) *t0* d(D(c)/dt * pybamm.grad(c)  
seems the function pybamm.Symbol.diff() does not work here.

**Reply by valentinsulzer** [reply, marked as answer]
It's best to refactor your equations so that the d/dt term is on the left-hand side and then add the rhs to `model.rhs`. If the d/dt term appears in multiple places then you can reuse the rhs expression. e.g. 
```python
da/dt = 2*a
db/dt = da/dt - b
```
becomes
```python
da/dt = 2*a
db/dt = 2*a - b
```
If you have a system that can't be rearranged in this way let us know

**Reply by xrfe** [reply]
That's probably the standard way to implement d/dt. in case I don't have explicit expression for the right hand side, I tried to use  pybamm.VariableDot(), e.g.,
j_n  = pybamm.Variables("Negative electrode interfacial current density [A.m-2]")
dj_ndt =  pybamm.VariableDot(j_n)
the error messege says the time derivative is on the right hand side.

**Reply by valentinsulzer** [reply]
Can you share the full system of equations? Number of equations should be equal to the number of variables, otherwise the system is not well defined

**Reply by xrfe** [reply]
besides the standard DFN equations, we have an additional equation for side reactions (li plating):
`d c_Li/ dt = - a*exp(-k*dj_n/dt)*N_Li`
the above equation quite resembles the default one in pybamm but add a coefficient function exp(k*dj_n/dt). so I have to access the derivative of the current density j_n, but it seems not able to directly use pybamm.VariableDot(j_n).

**Reply by brosaplanella** [reply]
If `j_n` is defined as `d j_n / dt = ...` then you can do what Valentin suggested above. If it is not, you can probably write `d j_n / dt` in terms of the other variables and apply the approach above.
